### PR TITLE
Raise ValueError when loss function is incompatible with activation function of output layer.

### PR DESCRIPTION
**Pull request by airalcorn2**
This is an alternative fix to #4744. I prefer this one.

### CODE REVIEW

**Comment by airalcorn2** [comment]
Many of the failing tests are actually additional evidence of the sneakiness of bug #4744. Let's take a closer look at the `test_masking_layer` test. 

```
import numpy as np
from numpy.testing import assert_allclose

from keras.layers import Dense, recurrent, TimeDistributed
from keras.models import Sequential
from keras.layers.core import Masking

from keras import backend as K

nb_samples, timesteps, embedding_dim, output_dim = 2, 5, 4, 3
embedding_num = 12

# @keras_test
# def test_masking_layer():
''' This test based on a previously failing issue here:
https://github.com/fchollet/keras/issues/1567

'''
I = np.random.random((6, 3, 4))
V = np.abs(np.random.random((6, 3, 5)))
V /= V.sum(axis=-1, keepdims=True)

model = Sequential()
model.add(Masking(input_shape=(3, 4)))
model.add(recurrent.LSTM(output_dim=5, return_sequences=True, unroll=False))
# model.add(TimeDistributed(Dense(output_dim = 1, activation='softmax'))) # This line should be here.
model.compile(loss='categorical_crossentropy', optimizer='adam')
model.predict(I)
```

Notice how the original code uses a `categorical_crossentropy` loss even though the output from a vanilla LSTM ranges from -1 to 1 because of the gated hyperbolic tangent activation function. Keras currently clips the output values before calculating the loss, so it's not clear what the model would be learning here.

For the `test_sequential_count_params` test, the code uses a `binary_crossentropy` loss even though the model is framed as a softmax over two output units. I think it would be a better practice to use the `categorical_crossentropy` in these situations.

```
import numpy as np

from keras.layers.core import Activation, Dense
from keras.models import Sequential

from keras import backend as K

# @keras_test
# def test_sequential_count_params():
input_dim = 20
nb_units = 10
nb_classes = 2

n = input_dim * nb_units + nb_units
n += nb_units * nb_units + nb_units
n += nb_units * nb_classes + nb_classes

model = Sequential()
model.add(Dense(nb_units, input_shape=(input_dim,)))
model.add(Dense(nb_units))
model.add(Dense(nb_classes))
model.add(Activation('softmax'))
model.build()

assert(n == model.count_params())

model.compile('sgd', 'binary_crossentropy')
# model.compile('sgd', 'categorical_crossentropy') # Should be this.
assert(n == model.count_params())
```

**Comment by fchollet** [comment]
This is far too restrictive. In particular `binary_crossentropy` should be usable with nearly any activation function (the only constraint being that the output should be between 0 and 1). E.g. it makes sense to use `binary_crossentropy` together with `softmax` (albeit it is unusual and generally suboptimal, depending on what you are trying to do).

**Comment by the-moliver** [comment]
I have mixed feelings about this. On the one hand I see the value of preventing incompatible output nonlinearity / loss function combos, on the other hand there are an endless number of naive mistakes one can make and it doesn't seem reasonable to expect Keras to catch them all.

**Comment by airalcorn2** [comment]
@the-moliver - I'm less interested in preventing "naive" mistakes and more interested in preventing unexpected behavior, e.g., a model continuing to "train" even though the pairing between the activation function on the output layer and the loss function doesn't make sense (e.g., a regression that's producing values far outside of zero and one being paired with a binary cross entropy loss). I think the "naive" mistake here would be to not clip your outputs when training with a, e.g., `binary_crossentropy` `loss`. With that in mind, I've implemented a `Clip` layer to give the user control over the clipping process.

**Comment by the-moliver** [comment]
I disagree with almost all of that. As @fchollet mentioned, by definition only values between 0 and 1 should be fed to `binary_crossentropy` loss. The best way to accomplish this is with a differentiable non-linearity such as a sigmoid, that maps all incoming values to 0->1. (Also, the `binary_crossentropy` loss-`sigmoid` nonlinearity is the cannonical pairing as in logistic regression, and should be the clear choice, unless one has a really good reason to do otherwise.) Clipping should only really be used for numerical stability, not to ensure the input to `binary_crossentropy`  is between 0 and 1. One _could_ do that, but your gradients are likely to get messed up, since anything in the clipped region will have 0 gradient. A sigmoid will give nice gradients over a much larger range (although not infinite range, hence the clipping). 
For those interested, a nice post on related issues is here: https://medium.com/@karpathy/yes-you-should-understand-backprop-e2f06eab496b#.lpox1bi3f

**Comment by airalcorn2** [comment]
@the-moliver: I should've clarified that the "naive" mistake would be to train a model that is using a sigmoid activation and a binary cross entropy loss and then not clipping the values to ensure numerical stability. I've edited the original comment to reflect that. I agree with your objections to using clipping to ensure the inputs to the `binary_crossentropy` loss lie between 0 and 1, but that's exactly what [Keras is currently doing](https://github.com/fchollet/keras/blob/5e75b8506c25f384311722f35b5dbbc3dd85f3fe/keras/backend/theano_backend.py#L1156) and it's the whole reason I opened issue #4744 in the first place.

**Comment by the-moliver** [comment]
Yes, and my point is that if you're doing things that make sense (like properly pairing loss functions and output nonlinearities), then there is no issue with the status quo. There's only aberrant behavior if you do things that don't make sense, like not mapping to 0-1 before using 'binary-crossentropy`, hence my original comment.

**Comment by fchollet** [comment]
Agreed, the current setup is fine. As @the-moliver points out, the current clipping happening with crossentropy exists for numeric stability: without it, a perfectly valid architecture would be very likely to end up with a NaN loss.
